stdlib/asm/crypto/hashes/rpo.masm

```diff
@@ -100,3 +100,152 @@ export.hash_memory
+#!
+#! Inputs:  [inputs_ptr, num_inputs]
+#! Outputs: [HASH]
+export.hash_memory
+    # move number of inputs to the top of the stack 
+    swap
+    # => [num_inputs, inputs_ptr]
+
+    # get the number of double words
+    u32divmod.8 swap
+    # => [num_inputs/8, num_inputs%8, inputs_ptr]
+
+    # get the end_addr for hash_memory_even proc (end address for pairs of words)
+    mul.2 dup.2 add movup.2
+    # => [inputs_ptr, end_addr, num_inputs%8]
+
+    # get the capacity element which is equal to num_inputs%8
+    dup.2
+    # => [capacity, inputs_ptr, end_addr, num_inputs%8]
+    
+    # prepare hasher state for RPO permutation 
+    push.0.0.0 padw padw 
+    # => [C, B, A, inputs_ptr, end_addr, num_inputs%8]
+
+    # hash every pair of words
+    exec.absorb_double_words_from_memory
+    # => [C', B', A', inputs_ptr', end_addr, num_inputs%8] where inputs_ptr' = end_addr
+
+    # hash remaining input values if there any left
+    # if num_inputs%8 is not ZERO
+    dup.14 push.0 neq
+    if.true
+        # load the remaining double word
+        mem_stream
+        # => [E, D, A', inputs_ptr'+2, end_addr, num_inputs%8]
+
+        # clean the stack
+        movup.12 drop movup.12 drop
+        # => [E, D, A', num_inputs%8]
+
+        # get the number of elements we need to drop
+        # notice that drop_counter could be any number from 1 to 7
+        push.8 movup.13 sub movdn.12
+        # => [E, D, A', drop_counter]
+
+        ### 0th value ########################################################
+        
+        # we need to drop first value anyway, since number of values is not divisible by 8
+        drop
+        # => [e_1, e_2, e_3, d_0, d_1, d_2, d_3, A', drop_counter]
+
+        # push the padding 0 value 
+        push.0
+        # => [0, e_1, e_2, e_3, d_0, d_1, d_2, d_3, A', drop_counter]
+
+        # move the calculated value down the stack 
+        movdn.6
+        # => [e_1, e_2, e_3, d_0, d_1, d_2, 0, d_3, A', drop_counter]
+
+        ### 1st value ########################################################
+
+        # prepare the second element of the E Word (e_1) for cdrop instruction
+        push.0 swap
+        # => [e_1, 0, e_2, e_3, d_0, d_1, d_2, 0, d_3, A', drop_counter]
+
+        # push latch variable onto the stack; this will be the control for the cdrop instruction
+        push.0
+        # => [latch = 0, e_1, 0, e_2, e_3, d_0, d_1, d_2, 0, d_3, A', drop_counter]
+
+        # get the flag whether the drop counter is equal 1
+        dup.14 eq.1
+        # => [drop_counter == 1, latch = 0, e_1, 0, e_2, e_3, d_0, d_1, d_2, 0, d_3, A', drop_counter]
+
+        # update the latch: if drop_counter == 1, latch will become 1
+        or
+        # => [latch', e_1, 0, e_2, e_3, d_0, d_1, d_2, 0, d_3, A', drop_counter]
+
+        # save the latch value 
+        dup movdn.14
+        # => [latch', e_1, 0, e_2, e_3, d_0, d_1, d_2, 0, d_3, A', latch', drop_counter]
```

Comment:
is there any reason as to why the latch needs to be this deep in the stack? why not on top of A'?

Reply:
There is no actual reason, just intuitive desire not to divide RPO words with some other values. We can store it before A' (in position 9), but on both these cases we need to emulate operations (`movdn.9` and `movdn.14`). We can although store latch before `d_3`, in that case we will use `movdn.8` without any emulations, but not sure that this optimization worth it: in that case tracking of elements in the stack could become a little bit confusing for anyone who wants to understand how does it work. 

But I don't have any preferences, so maybe we indeed should store latch before `d_3`.